### ISSUE TITLE
[Bolt] issue about using std::hash in YamlProfileReader and its test case

### ISSUE THREAD

**Issue by linsinan1995** [llvm/llvm-project#65241]
bolt tries to match block counts based on hashes generated from std::hash. However, std::hash does not guarantee the result will be the same from different setups.

I recently used a llvm-bolt linked against libc++, and it failed on test case X86/reader-stale-yaml.test, since the std::hash results are different from these two libraries...

<img width="505" alt="image" src="https://github.com/llvm/llvm-project/assets/47880367/6e7dd792-37ba-48d7-80c8-6b84897f3a94">

**Comment by aaupov** [comment]
Thanks for bringing this up. We should fix this since it fails with libcxx, perhaps by switching to xxh3 just like in couple other places in llvm.

**Comment by linsinan1995** [comment]
> Thanks for bringing this up. We should fix this since it fails with libcxx, perhaps by switching to xxh3 just like in couple other places in llvm.

Hi @aaupov , it looks like a sound solution, and it would be nicer if you could also fix the other two std::hash usages in `bolt/lib/Profile/StaleProfileMatching.cpp` :)

**Comment by aaupov** [comment]
Yes, but modifying stale matching hashes creates a problem that checked in tests no longer match the intended way. Leaving the issue open.
